File: flutter-app/lib/models/room.dart
class Room {
  final String id;
  final String name;
  final String? description;
  final bool isPublic;
  final String creatorId;
  final String? roomKey;
  final int maxMembers;
  final DateTime createdAt;
  final int? memberCount;
  final int? messageCount;
  int unreadCount;

  Room({
    required this.id,
    required this.name,
    this.description,
    required this.isPublic,
    required this.creatorId,
    this.roomKey,
    this.maxMembers = 100,
    required this.createdAt,
    this.memberCount,
    this.messageCount,
    this.unreadCount = 0,
  });

  factory Room.fromJson(Map<String, dynamic> json) {
    return Room(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      isPublic: json['isPublic'] as bool,
      creatorId: json['creatorId'] as String,
      roomKey: json['roomKey'] as String?,
      maxMembers: json['maxMembers'] as int? ?? 100,
      createdAt: DateTime.parse(json['createdAt'] as String),
      memberCount: json['memberCount'] as int?,
      messageCount: json['messageCount'] as int?,
      unreadCount: json['unreadCount'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'isPublic': isPublic,
      'creatorId': creatorId,
      'roomKey': roomKey,
      'maxMembers': maxMembers,
      'createdAt': createdAt.toIso8601String(),
      'memberCount': memberCount,
      'messageCount': messageCount,
      'unreadCount': unreadCount,
    };
  }

  Room copyWith({
    String? id,
    String? name,
    String? description,
    bool? isPublic,
    String? creatorId,
    String? roomKey,
    int? maxMembers,
    DateTime? createdAt,
    int? memberCount,
    int? messageCount,
    int? unreadCount,
  }) {
    return Room(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      isPublic: isPublic ?? this.isPublic,
      creatorId: creatorId ?? this.creatorId,
      roomKey: roomKey ?? this.roomKey,
      maxMembers: maxMembers ?? this.maxMembers,
      createdAt: createdAt ?? this.createdAt,
      memberCount: memberCount ?? this.memberCount,
      messageCount: messageCount ?? this.messageCount,
      unreadCount: unreadCount ?? this.unreadCount,
    );
  }
}

File: rust-backend/src/models/room.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Room {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    #[serde(rename = "type")]
    #[sqlx(rename = "type")]
    pub room_type: String,
    pub encryption_key: String,
    pub creator_id: Option<Uuid>,
    pub max_members: i32,
    pub is_public: bool,
    pub avatar: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct CreateRoomRequest {
    #[validate(length(min = 1, max = 100))]
    pub name: String,

    #[validate(length(max = 500))]
    pub description: Option<String>,

    #[serde(rename = "type")]
    pub room_type: Option<String>,

    #[validate(range(min = 2, max = 1000))]
    pub max_members: Option<i32>,

    pub is_public: Option<bool>,

    pub avatar: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RoomResponse {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    #[serde(rename = "type")]
    pub room_type: String,
    pub room_key: Option<String>,
    pub creator_id: Option<Uuid>,
    pub max_members: i32,
    pub is_public: bool,
    pub avatar: Option<String>,
    pub created_at: DateTime<Utc>,
}

impl Room {
    pub fn to_public_json(&self) -> RoomResponse {
        RoomResponse {
            id: self.id,
            name: self.name.clone(),
            description: self.description.clone(),
            room_type: self.room_type.clone(),
            room_key: None,
            creator_id: self.creator_id,
            max_members: self.max_members,
            is_public: self.is_public,
            avatar: self.avatar.clone(),
            created_at: self.created_at,
        }
    }

    pub fn to_member_json(&self) -> RoomResponse {
        RoomResponse {
            id: self.id,
            name: self.name.clone(),
            description: self.description.clone(),
            room_type: self.room_type.clone(),
            room_key: Some(self.encryption_key.clone()),
            creator_id: self.creator_id,
            max_members: self.max_members,
            is_public: self.is_public,
            avatar: self.avatar.clone(),
            created_at: self.created_at,
        }
    }
}
